trieste/models/gpflux/optimizer.py

```diff
@@ -0,0 +1,49 @@
+# Copyright 2021 The Trieste Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import tensorflow as tf
+from gpflux.models import DeepGP
+from tensorflow.python.data.ops.iterator_ops import OwnedIterator as DatasetOwnedIterator
+
+from ..optimizer import LossClosure, TrainingData, create_loss_function
+
+
+@create_loss_function.register
+def _create_loss_function_gpflux(
```

Comment:
GPflux is using `fit` method from Keras to train DGPs, `fit` provides a lot functionality that we don't want to replicate - `TFOptimizer` will probably work ok, but is not ideal e.g. we create batches there (or through `DatasetTransformer`), I would think Keras' `fit` is more reliable

perhaps we can postpone it for later, but this would require another type of `Optimizer` class, it can be Keras specific or agnostic, such that we can use training method provided by the library, rather than calling an optimizer directly

Reply:
I have had a go at this, but not by making a new `Optimizer` class, because that seems to run into issues with dispatching to the wrong `create_optimizer` function, as `tf.optimizers.Optimizer` is always a `tf.keras.optimizers.Optimizer`, so that it dispatches to create a `KerasOptimizer` even when we want to create  `TFOptimizer` (for instance, for SVGP). At the moment I don't really see an easy way around this.

My solution was to instead just use the optimizer from `TFOptimizer`, along with the args passed to it, to use `fit` when `model.optimize(Dataset)` is called. See line 98 in `gpflux/interface.py`.